serverless/cmd/integrate/main.go

```diff
@@ -37,8 +44,60 @@ var (
 	pubKeyFile  = flag.String("public_key", "", "Location of public key file. If unset, uses the contents of the SERVERLESS_LOG_PUBLIC_KEY environment variable.")
 	privKeyFile = flag.String("private_key", "", "Location of private key file. If unset, uses the contents of the SERVERLESS_LOG_PRIVATE_KEY environment variable.")
 	origin      = flag.String("origin", "", "Log origin string to use in produced checkpoint.")
+	dsmKeyId    = flag.String("dsm_key_id", "", "ID of key in DSM to use for signing")
+	dsmApiKey   = flag.String("dsm_api_key", "", "API key for accessing smartkey")
+	dsmEndpoint = flag.String("dsm_endpoint", "https://www.smartkey.io", "DSM instance to use when signing with DSM")
+	pubKeyOut   = flag.String("public_key_out", "", "When initializing and using a DSM signing key, write the public key to this file")
 )
 
+const Ed25519PubKeyLength = 32
+
+// The keyHash implementation is copied from the Note keyHash implementation. This function is private to Note
+// for no good reason. We want to use the same key hashing function because the key hashes are used internally
+// by the Note package for identifying which keys were used to sign a Note.
+
+// keyHash computes the key hash for the given server name and encoded public key.
+func keyHash(name string, key []byte) uint32 {
+	h := sha256.New()
+	h.Write([]byte(name))
+	h.Write([]byte("\n"))
+	h.Write(key)
+	sum := h.Sum(nil)
+	return binary.BigEndian.Uint32(sum)
+}
+
+// This type and associated functions implement the note.Signer signing interface, but sign using a key in DSM.
+type dsmSigner struct {
+	name string
+	hash uint32
+	client *sdkms.Client
+	key *sdkms.SobjectDescriptor
+}
+
+func (signer *dsmSigner) Name() string {
```

Comment:
I couldn't understand why you are creating such a function.

Reply:
Name, KeyHash, and Sign are the required methods for the Signer type, which is part of the golang Note package. https://pkg.go.dev/golang.org/x/mod/sumdb/note#Signer

The Trillian serverless tooling expects to be passed a Signer, and it calls these interfaces to perform the signing. The Note data format includes the key's name and keyhash along with the signature, to help signature verifiers find the correct key for verifying the signature.